model_analyzer/entrypoint.py

```diff
@@ -0,0 +1,210 @@
+from .monitor.nvml import NVMLMonitor
+from .perf_analyzer.perf_analyzer import PerfAnalyzer
+from .perf_analyzer.perf_config import PerfAnalyzerConfig
+from .triton.server.server_factory import TritonServerFactory
+from .triton.server.server_config import TritonServerConfig
+from .triton.client.client_factory import TritonClientFactory
+from .triton.client.client_config import TritonClientConfig
+from .triton.model.model import Model
+from .record.record_aggregator import RecordAggregator
+from .record.gpu_free_memory import GPUFreeMemory
+from .record.gpu_used_memory import GPUUsedMemory
+from .record.perf_throughput import PerfThroughput
+from .output.output_table import OutputTable
+from .output.file_writer import FileWriter
+
+
+def create_perf_config(params):
+    """
+    Utility function for creating
+    a PerfAnalyzerConfig from
+    a dict of parameters
+
+    Parameters
+    ----------
+    params : dict
+        keys are arguments to perf analyzer
+        values are their value
+
+    Returns
+    -------
+    PerfAnalyzerConfig
+        The arguments from params
+        are set in this config
+    """
+
+    config = PerfAnalyzerConfig()
+    for param, value in params.items():
+        config[param] = value
+    return config
+
+
+def main():
+    """
+    Main entrypoint of model_analyzer
+    """
+
+    # Get args
+    cli = CLI()
+    args = cli.parse(sys.argv)
+    print(args)
+
+    # Create and start tritonserver
+    triton_config = TritonServerConfig()
+    triton_config['model-repository'] = args.model_repository
+    triton_config['model-control-mode'] = 'explicit'
+
+    server = TritonServerFactory.create_server_local(config=triton_config)
+    server.start()
+    server.wait_for_ready()
+
+    # Create triton client and model
+    client_config = TritonClientConfig()
+    if args.client_protocol == 'http':
+        client_config['url'] = 'localhost:8000'
+        client = TritonClientFactory.create_http_client(config=client_config)
+    elif args.client_protocol == 'grpc':
+        client_config['url'] = 'localhost:8001'
+        client = TritonClientFactory.create_grpc_client(config=client_config)
+
+    # To run perf_analyzer we need all combinations of configs
+    start, stop, step = tuple(map(int, args.concurrency_range.split(':')))
+    sweep_params = {
+        'model-name': args.model_names.split(','),
+        'batch-size': args.batch_size.split(','),
+        'concurrency-range': list(range(start, stop + 1, step))
+    }
+
+    param_combinations = list(product(*tuple(sweep_params.values())))
+    run_params = [
+        dict(zip(sweep_params.keys(), vals)) for vals in param_combinations
+    ]
+
+    # Create a record aggregator to collect records, and output table
+    record_aggregator = RecordAggregator()
+    table_headers = [
+        "Model", "Batch", "Concurrency",
+        PerfThroughput.header(), "Max " + GPUFreeMemory.header(),
```

Comment:
Can we refactor this to make it more generalizable? For example, the whole metrics that are being collected depends on a list that we can simply add/remove to it to change the metrics being monitored.

```
monitoring_metrics = [GPUFreeMemory, GPUUsedMemory, GPUUtilization]

# The rest of the code will depend on this array...
```